### DISCUSSION TITLE
Add an option to exlcude basic models from mappings

### DISCUSSION THREAD

**Discussion by heindrickdumdum0217** [wallneradam/esorm#13] [Q&A] [answered]
Environment
```
Pydantic == 2.8.2
pyesorm == 0.4.8
```

I have several models, and they have common fields.
So I created a base model, but when call `setup_mappings` function, it tries to create an index for the base model too.

```
from esorm.model import ESModel


class _BaseModel(ESModel):
    id: unsigned_long


class A(_BaseModel):
    name: str
```

It throws the following error.
```
Index mappings error:
Traceback (most recent call last):
File \"/home/xxx/.pyenv/versions/3.11.5/envs/py3.11.5/lib/python3.11/site-packages/esorm/model.py\", line 1207, in setup_mappings
await es.indices.create(index=index,\n  File \"/home/xxx/.pyenv/versions/3.11.5/envs/py3.11.5/lib/python3.11/site-packages/elasticsearch/_async/client/indices.py\", line 553, in create
return await self.perform_request(  # type: ignore[return-value]     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/home/xxx/.pyenv/versions/3.11.5/envs/py3.11.5/lib/python3.11/site-packages/elasticsearch/_async/client/_base.py\", line 423, in perform_request
return await self._client.perform_request(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/home/xxx/.pyenv/versions/3.11.5/envs/py3.11.5/lib/python3.11/site-packages/elasticsearch/_async/client/_base.py\", line 271, in perform_request
response = await self._perform_request(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File \"/home/xxx/.pyenv/versions/3.11.5/envs/py3.11.5/lib/python3.11/site-packages/elasticsearch/_async/client/_base.py\", line 352, in _perform_request
raise HTTP_EXCEPTIONS.get(meta.status, ApiError)(
elasticsearch.BadRequestError: BadRequestError(400, 'invalid_index_name_exception', 'Invalid index name [myproject__BaseModel], must be lowercase')
```

The error doesn't affect to other indices, but I think still need an option to ignore some models from mappings.
For example:
```
class ESConfig:
    abstract = True
```

**Comment by wallneradam** [comment]
The [ESBaseModel](https://github.com/wallneradam/esorm/tree/main?tab=readme-ov-file#esbasemodel) is exactly for this.

**Comment by wallneradam** [comment]
Oh, I misunderstood your question...
I'm going to test if the following works:
```python
import esorm


class BaseModel(esorm.ESBaseModel):
    ...


class Model(BaseModel, esorm.ESModel):
    ...
```

I'm not sure if it works atm. but if not I will make it work this way.

**Comment by heindrickdumdum0217** [comment]
@wallneradam 

Yes, we can use two base models.
But I prefer to use only one base model and can we add `Config` class to `ESBaseModel`?
For all models, I want to add an option to ignore extra.
```
class _BaseModel(ESModel):
    id: unsigned_long

    class Config:
        extra = Extra.ignore
```

**Comment by wallneradam** [comment, marked as answer]
Ok, I've tested it, this works, and it is more pythonic way. You have one base model, but two parents:

```python
class BaseModelParent(esorm.ESBaseModel):
    class ESConfig:
        id_field = 'f_str'

    f_str: str
    f_int: int

class BaseModelParentModel(BaseModelParent, esorm.ESModel):
    f_float: float
```

`ESBaseModel` classes won't be incuded in mappings. So you can use them as a base structure for multiple models. Also `ESConfig` settings are inherited from the base model. I think this is what yout want.

**Comment by heindrickdumdum0217** [comment]
@wallneradam 
I just tested, parent model config is not extended to child model.

```
class _BaseModel(ESBaseModel):
    id: unsigned_long

    class Config:
        extra = Extra.ignore


class A(_BaseModel, ESBaseModel)
    name: str


data = {"id": 3, "name": "Bob", "age": 17}
a = A.parse_obj(data)
await a.save()
```

It throws the following error.
```
Extra inputs are not permitted
```

I'm just giving an example.
In real situation data is from third party and it has very complex structure, this is why I'm using `parse_obj` function.

** Note **
I'm saying about `Config` class not `ESConfig` class.

**Comment by wallneradam** [comment]
1. There's no `Config` class in `Pydantic` [v2.x](https://docs.pydantic.dev/latest/migration/#changes-to-config). After that it will be inherited according to the [documentation](https://docs.pydantic.dev/2.0/usage/model_config/).
2. If you change `extra` to `allow`, i'm not sure it will work as intended. At least it is not tested. I don't think is a good idea to change `model_config` at all in Esorm models, it is fine tuned for ElasticSearch.
3. Why you would like to add extra? Esorm is for easily make "schema" for your model. Which can be used everywhere in the IDE and it is optimized in the DB. If you can add extra fields, the main goal will hurt.

**Comment by heindrickdumdum0217** [comment]
@wallneradam 
`ESBaseModel` has `model_config`, I can't overwride the existing config.

https://github.com/wallneradam/esorm/blob/main/esorm/model.py#L237-L244

**Reply by wallneradam** [reply]
You can. I linked the documentation about it. Search this text: "If you add a model_config to the Model class, it will merge with the model_config from Parent".

It is the magic of Pydantic. I use similar magic with ESConfig class.
